File: all-permutations.cpp
#include <iostream>
#include <random>
#include <unordered_map>
#include <vector>
#include <algorithm>
#include <unordered_set>

using namespace std;

random_device rd;
mt19937 gen(rd());

long long random(long long lower_bound, long long upper_bound) {
    uniform_int_distribution<long long> dist(lower_bound, upper_bound);
    return dist(gen);
}

long long factorial(long long number) {
    long long total = 1;

    while (number != 1) {
        total *= number;
        number -= 1;
    }
    return total;
}

int main() {
    string word;

    while (true) {
        cout << "Enter a word: ";
        getline(cin, word);

        if (word.empty()) {
            cout << "Word must not be left empty.\n" << endl;
        }

        else {
            break;
        }
    }

    // Remove trailing and leading spaces from word later
    unordered_set<string> permutations = {word};
    unordered_map<char, int> letters;
    vector<char> unique_keys = {};
    long long denominator = 1;

    for (char letter: word) {
        if (letters.count(letter) == 0) {
            letters[letter] = 1;
            unique_keys.push_back(letter);
        }
        else {
            letters[letter] += 1;
        }
    }

    for (auto value: letters) {
        denominator *= factorial(value.second);
    }

    long long no_of_combinations = factorial(word.length()) / denominator;

    cout << "This word has " + to_string(no_of_combinations) + " combination(s)." << endl;


    while (permutations.size() < no_of_combinations) {
        unordered_map<char, int> letters_copy = letters;
        vector<char> unique_keys_copy = unique_keys;
        vector<char> generated_vector = {};
        string generated_word;
        vector<int> indexes = {};
        int random_index = 0;
        char random_letter = '\0';
        int number_pos = 0;
        int letter_pos = 0;

        for (int i = 0; i < word.length(); i++) {
            generated_vector.push_back('\0');
            indexes.push_back(i);
        }

        for (int counter = 0; counter < word.length(); counter++) {

            number_pos = random(0, (indexes.size() - 1));
            random_index = indexes.at(number_pos);
            indexes.erase(indexes.begin() + number_pos);

            letter_pos = random(0, (unique_keys_copy.size() - 1));
            random_letter = unique_keys_copy.at(letter_pos);
            letters_copy.at(random_letter) -= 1;

            if (letters_copy.at(random_letter) == 0) {
                letters_copy.erase(random_letter);
                unique_keys_copy.erase(unique_keys_copy.begin() + letter_pos);
            }

            generated_vector.at(random_index) = random_letter;
        }

        for (char letter : generated_vector) {
            generated_word += letter;
        }

        permutations.insert(generated_word);
    }

    cout << "All permutations: " << endl;

    for (string word: permutations) {
        cout << word + ", ";
    }
}

File: all_permutations.java
import java.util.*;

public class all_permutations {

    public static long factorial(long number) {
        long total  = 1;

        while (number != 1) {
            total *= number;
            number -= 1;
        }

        return total;
    }

    public static void main(String[] args) {
        String word;
        Scanner input = new Scanner(System.in);
        Random random = new Random();

        System.out.println(System.lineSeparator() + "\u001B[4mRejection Sampling Permutation Generator\u001B[0m");

        while (true) {
            System.out.print(System.lineSeparator() + "Enter a word (all characters are acceptable): ");
            word = input.nextLine();

            if (word.isEmpty()) {
                System.out.println(System.lineSeparator() + "Word must not be left empty.");
            }
            else {
                break;
            }
        }

        input.close();

        HashSet<String> permutations = new HashSet<>();
        HashMap<Character, Integer> letters = new HashMap<>();
        ArrayList<Character> uniqueKeys = new ArrayList<>();
        long denominator = 1;

        for (char letter : word.toCharArray()) {
            if (!letters.containsKey(letter)) {
                letters.put(letter, 1);
                uniqueKeys.add(letter);
            }
            else {
                letters.put(letter, letters.get(letter) + 1);
            }
        }

        for (int value: letters.values()) {
            denominator *= factorial(value);
        }

        long numOfCombinations = factorial(word.length()) /  denominator;

        System.out.println(System.lineSeparator() + "This word has " +  numOfCombinations + " combination(s).");

        while (permutations.size() < numOfCombinations) {
            HashMap<Character, Integer> lettersCopy = (HashMap<Character, Integer>) letters.clone();
            List<Character> uniqueKeysCopy = (ArrayList<Character>) uniqueKeys.clone();
            List<Character> generatedList = new ArrayList<>();
            StringBuilder generatedWord = new StringBuilder();
            List<Integer> indexes = new ArrayList<>();
            int randomIndex;
            char randomLetter;
            int numberPosition;
            int letterPosition;

            for (int i = 0; i < word.length(); i++) {
                generatedList.add(null);
                indexes.add(i);
            }

            for (int counter = 0; counter < word.length(); counter++) {

                numberPosition = random.nextInt(indexes.size());
                randomIndex = indexes.get(numberPosition);
                indexes.remove(numberPosition);

                letterPosition = random.nextInt(uniqueKeysCopy.size());
                randomLetter = uniqueKeysCopy.get(letterPosition);
                lettersCopy.put(randomLetter, lettersCopy.get(randomLetter) - 1);

                if (lettersCopy.get(randomLetter) == 0) {
                    lettersCopy.remove(randomLetter);
                    uniqueKeysCopy.remove(letterPosition);
                }

                generatedList.set(randomIndex, randomLetter);
            }

            for (char letter: generatedList) {
                generatedWord.append(letter);
            }

            permutations.add(generatedWord.toString());
        }

        System.out.println(System.lineSeparator() + "All permutations: " + System.lineSeparator() + permutations);
    }
}